Dalamud/Interface/Internal/Windows/Data/Widgets/UldWidget.cs

```diff
@@ -0,0 +1,1210 @@
+                break;
+            case Bool3Keyframe bool3Keyframe:
+                this.DrawTimelineKeyGroupFrame(bool3Keyframe.Keyframe);
+                ImGui.SameLine(0, 0);
+                ImGui.TextUnformatted($" | Value1: {bool3Keyframe.Value[0]} | Value2: {bool3Keyframe.Value[1]} | Value3: {bool3Keyframe.Value[2]}");
+                break;
+            case ColorKeyframe colorKeyframe:
+                this.DrawTimelineKeyGroupFrame(colorKeyframe.Keyframe);
+                ImGui.SameLine(0, 0);
+                ImGui.TextUnformatted($" | Add: {colorKeyframe.AddRed} {colorKeyframe.AddGreen} {colorKeyframe.AddBlue} | Multiply: {colorKeyframe.MultiplyRed} {colorKeyframe.MultiplyGreen} {colorKeyframe.MultiplyBlue}");
+                break;
+            case LabelKeyframe labelKeyframe:
+                this.DrawTimelineKeyGroupFrame(labelKeyframe.Keyframe);
+                ImGui.SameLine(0, 0);
+                ImGui.TextUnformatted($" | LabelCommand: {labelKeyframe.LabelCommand} | JumpId: {labelKeyframe.JumpId} | LabelId: {labelKeyframe.LabelId}");
+                break;
+        }
+    }
+
+    private unsafe void DrawParts(UldRoot.PartsData partsData, UldRoot.TextureEntry[] textureEntries, DataManager dataManager, TextureManager textureManager)
+    {
+        if (ImGui.CollapsingHeader($"Parts {partsData.Id}"))
+        {
+            for (var index = 0; index < partsData.Parts.Length; index++)
+            {
+                var partsDataPart = partsData.Parts[index];
+                var texturePathChars = textureEntries.First(t => t.Id == partsDataPart.TextureId).Path;
+                string texturePath;
+                fixed (char* p = texturePathChars)
+                    texturePath = new string(p);
+                var texFile = dataManager.GetFile<TexFile>(texturePath);
+                if (texFile == null) continue;
+                var wrap = textureManager.CreateFromTexFile(texFile);
+                var texSize = new Vector2(texFile.Header.Width, texFile.Header.Height);
+                var uv0 = new Vector2(partsDataPart.U, partsDataPart.V);
+                var partSize = new Vector2(partsDataPart.W, partsDataPart.H);
+                var uv1 = uv0 + partSize;
+                ImGui.TextUnformatted($"Index: {index}");
+                ImGui.SameLine();
+                ImGui.Image(wrap.ImGuiHandle, partSize, uv0 / texSize, uv1 / texSize);
+                wrap.Dispose();
+            }
+        }
+    }
+}
+
+/// <summary>
+/// Contains the raw data for the ULD widget.
+/// </summary>
+internal class UldWidgetData
+{
+    /// <summary>
+    /// All known uld locations referenced in executable.
+    /// Generated from FFXIVClientStructs\ida\ffxiv_extract_uld_names.py as of 7.01
+    /// </summary>
+    #pragma warning disable SA1401 // Fields should be private - this needs to be internal or public for above class to access it
+    internal static string[] UldLocations =
```

Comment:
This will need to be generated and manually updated on a per-patch cycle?

I assume there's no automatic way to get it, nor any better place this can go then...

Reply:
> This will need to be generated and manually updated on a per-patch cycle?
Yesn't... Depends what kind of patch it is since they might not add uld files but instead change said file with new info.

> I assume there's no automatic way to get it, nor any better place this can go then...
I didn't fully look into how to see if I could get data from the asset path that say `bannedplugins` is loaded from, as it could be also located there. 

I will note that the list that said script currently generate isn't complete, as there are certain ones that it can find by sig scanning.

Reply:
As-is, I'm a bit wary of having a list that we need to update anywhere inside Dalamud, especially for something non-critical like a developer system. Any changes to this list risk going unnoticed for a while.

Moving forward, I think we'd need to either:

1. Add a plain text field to allow developers to import raw ULD names (this is probably a good idea anyways)
2. Add some sort of smarter auto-detection logic to find ULDs. The game ultimately needs to know about all the different ULD kinds, so we should be able to use the same systems to find things.

We've discussed point 2 in the Discord a bit, and it seems harder than expected as there's no good way to match for `ui/uld/...` in the EXE. Most ULDs are loaded via a call to `AtkUnitBase#LoadUldByName`, where the passed value comes from a series of constant strings in the binary that don't even have the politeness to be a proper array.

Reply:
I think I found a nice way to do the second one for ULD detection, with modifying the `SigScanner` a tiny bit to allow getting every match in the `.text` area of the game memory. This possibly will break if people modify said area, but not entirely sure.